### PR TITLE
check whether serviceCIDR contains clusterCIDR during ipam initialization

### PR DESCRIPTION
**Pull request by gongguan**
**What type of PR is this?**

**What this PR does / why we need it**:
When I started kube-controller-manager with option `--cluster-cidr=10.0.0.0/24 --service-cluster-ip-range=10.0.0.0/21`. 
while service-cluster-ip contains cluster-cidr, each node can't be assigned with podCIDR(cidr is 
all occupied by service ip) even controller-manager running with Error 
`"failed to allocate cidr: CIDR allocation failed; there are no remaining CIDRs left to allocate in the accepted range"`
 I think whether clusterCIDR is contained by service-CIDR should be checked during ipam controller initailization.

**Which issue(s) this PR fixes**:

Fixes # 

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 1]
```diff
@@ -379,3 +380,34 @@ func cidrsAsString(inCIDRs []*net.IPNet) []string {
 	}
 	return outCIDRs
 }
+
+// used to check whether a cidr contains another
+func containsCIDR(largeCIDR *net.IPNet, smallCIDR *net.IPNet) bool {
```
Comment (bowei) [inline pkg/controller/nodeipam/ipam/range_allocator.go]: Does this function not already exist somewhere?

File: pkg/controller/nodeipam/ipam/controller.go [hunk 2]
```diff
@@ -71,6 +71,10 @@ func NewController(
 		return nil, fmt.Errorf("invalid IPAM controller mode %q", config.Mode)
 	}
 
+	if containsCIDR(serviceCIDR, clusterCIDR) {
+		return nil, fmt.Errorf("serviceCIDR contains clusterCIDR, node's podCIDR can't be assigned")
+	}
+
```
Comment (bowei) [inline pkg/controller/nodeipam/ipam/controller.go]: seems like what would be more generic is to check that after occupyServiceCIDR, we have free space left in the clusterCIDR

**Review by gongguan — COMMENTED** [review]
update

**Comment by gongguan** [comment]
cc please @bowei

**Comment by gongguan** [comment]
@bowei redundant code removed

**Comment by gongguan** [comment]
/assign @bowei
redundant code removed

File: pkg/controller/nodeipam/ipam/controller.go [hunk 3]
```diff
@@ -92,6 +92,9 @@ func NewController(
 		return nil, err
 	}
 
+	if _, err := c.set.AllocateNext(); err != nil {
```
Comment (bowei) [inline pkg/controller/nodeipam/ipam/controller.go]: don't you need to dealloc

Comment (bowei) [reply pkg/controller/nodeipam/ipam/controller.go]: Also, there should be log with error and I would return a new error type.

Comment (gongguan) [reply pkg/controller/nodeipam/ipam/controller.go]: Done. There is no error log in the NewController function. Can I return a new error type without printing log?

**Comment by gongguan** [comment]
ping @bowei

File: pkg/controller/nodeipam/ipam/controller.go [hunk 4]
```diff
@@ -92,6 +92,10 @@ func NewController(
 		return nil, err
 	}
 
+	if _, err := c.set.ProbeCIDRLeft(); err != nil {
```
Comment (bowei) [inline pkg/controller/nodeipam/ipam/controller.go]: why not

```golang
cidr, err := c.set.AllocateNext()
switch err {
  case ErrCIDRRangeNoCIDRsRemaining:
    return nil, fmt.Errorf("no free IP space for pods in the CIDR range")
  case nil:
    c.set.Release(cidr)
    return c, nil
  default:
    return nil, fmt.Error("error when testing for free CIDR range: %v", err)
}
```

Comment (bowei) [reply pkg/controller/nodeipam/ipam/controller.go]: You don't have to add all of the other code.

Comment (gongguan) [reply pkg/controller/nodeipam/ipam/controller.go]: Done. In my perspective, check there is a space is more reasonable than allocate and release a cidr.

**Pull request merged**